Sources/MongoSwift/MongoConnectionString.swift

```diff
@@ -0,0 +1,373 @@
+import Foundation
+
+/// Represents a MongoDB connection string.
+/// - SeeAlso: https://docs.mongodb.com/manual/reference/connection-string/
+public struct MongoConnectionString: Codable, LosslessStringConvertible {
+    private static let forbiddenDBCharacters = ["/", "\\", " ", "\"", "$"]
+    /// Forbidden characters per RFC 3986.
+    /// - SeeAlso: https://datatracker.ietf.org/doc/html/rfc3986#section-2.2
+    fileprivate static let forbiddenUserInfoCharacters = [":", "/", "?", "#", "[", "]", "@"]
+
+    /// Represents a connection string scheme.
+    public struct Scheme: LosslessStringConvertible, Equatable {
+        /// Indicates that this connection string uses the scheme `mongodb`.
+        public static let mongodb = Scheme(.mongodb)
+
+        /// Indicates that this connection string uses the scheme `mongodb+srv`.
+        public static let srv = Scheme(.srv)
+
+        /// Internal representation of a scheme.
+        private enum _Scheme: String {
+            case mongodb
+            case srv = "mongodb+srv"
+        }
+
+        private let _scheme: _Scheme
+
+        private init(_ value: _Scheme) {
+            self._scheme = value
+        }
+
+        /// LosslessStringConvertible` protocol requirements
+        public init?(_ description: String) {
+            guard let _scheme = _Scheme(rawValue: description) else {
+                return nil
+            }
+            self.init(_scheme)
+        }
+
+        public var description: String { self._scheme.rawValue }
+    }
+
+    /// A struct representing a host identifier, consisting of a host and an optional port.
+    /// In standard connection strings, this describes the address of a mongod or mongos to connect to.
+    /// In mongodb+srv connection strings, this describes a DNS name to be queried for SRV and TXT records.
+    public struct HostIdentifier: Equatable, CustomStringConvertible {
```

Comment:
Why is a new type needed instead of `ServerAddress`?

Reply:
Actually not sure, this is part of the first commit which is just a reapply of Bynn's original commit to introduce `MongoConnectionString` -- was this design discussed during the original implementation?

Reply:
I remember we talked about this and ended up concluding we needed a new type but I do not remember why. looking at the differences though, the `port` is optional here but non-optional for `ServerAddress`, and this type also supports a host type whereas the other does not. I think the problem might have been that we need to be able to store unresolved DNS names here, whereas we could only store resolved ones in a `ServerAddress`?

Reply:
ah yeah that sounds familiar.